### PR TITLE
[41341] Updated date picker drop modal (including duration and non-working days)

### PR DESCRIPTION
**Pull request by HDinger**
## Acceptance criteria

**Duration**

- [x] There is a new "Duration" field to the right of finish date
- [x] The visual format for displaying duration is "4 days".
  - [x] When in focus, this retains only the numerical value "4".
  - [x] The user can only enter numbers in the duration field
  - [x] On blur (remove focus), the localised word  for "days" is added again.
- [x] When the Duration field is clicked on, it is considered "active"
- [x] Clicking outside of the duration field (blur) does not change the active field; it remains duration, and only changes:
  - [x] with a click on the start or finish date fields
  - [x] with a click on a date in the mini calendar

**Mini calendar updates**

- [x] In the selected range but a non-working day:
  - [x] When calendar is enabled: blue text on a gray#6 background
- [x] When calendar is disabled: gray#5 on a gray#6 background

**"Include non-working days" checkbox**

- [x] The action bar has a new checkbox on the left edge:
- [x] checking removes the grey background on non-working days; all days now look the same as a "normal" day.
- [x] Unchecking the box reverses this (and non-working days become grey again and are not selectable in the calendar).

**Data cleanup**
 
* Will be done in a separate PR

**Logical changes**
- [x] Changing either a date, the duration or the non-working days will have automatic effects on the other fields.The concrete logic is described in [#43539](https://community.openproject.org/work_packages/43539)
- [x] This calculation is done via the POST endpoint (see [#43651](https://community.openproject.org/projects/openproject/work_packages/43651/activity)

**Tests**

- [x] Add test for all scenarios defined [here](https://community.openproject.org/projects/openproject/work_packages/43539/activity) 
- [x] Add test for focusing the duration field when opening it from the table (see #11170)
- [x] Additional examples taken from the WP description:
  - [x] If a work package already has start and end dates (and thus also duration) and the user enables this checkbox, the duration is preserved the dates are updated to include days that were previously skipped (the weekends, primarily).
  - [x] If the user goes back to working days only (from previously included non-working days), the duration is once again maintained, but the finish date is pushed to now skip over the non-working days. (scenario 15+16)
  - [x] If a work package only has one set of date information (just start date, just end date, or just duration), ~~clicking this button will not affect any of the values until a second piece of information is also input.~~ toggling include NWD will move the single set date
  - [x] When a user modifies the duration integer:
    - [x] If no dates are entered, there is no impact on the mini-calendar. (scenario 2)
    - [x] If both dates are present, the finish date is updated automatically based on the input (preserved, or shifted into the past or the future) (scenario 4)
    - [x] If only a start date was present, a finish date is now derived. (scenario 1)
    - [x] If only a finish date was present, a start date is now derived. (scenario 3)
    - [x] When the user is in the duration field and clicks on a date in the mini calendar:
      - [x] ~~If the duration field has a value, the clicked date becomes the the start date, the finish date is derived from the duration, and the active field moves to the finish date (same as when a date is clicked on when the start date is active).~~ (superseded by scenario 13)
        - [x] A click on a second date then modifies the finish date (and can affect duration). The active field shifts again back to the start date.
      - [x] If the duration field has no value, the clicked date becomes the start date, no finish date is set but the active field becomes the finish date.
        - [x] A click on a second date modifies the finish date (which derives duration too). The active field shifts again back to the start date.

**Known issues**

- [x] Updating dates in the datepicker causes two form requests, one from datepicker and from the debounced change.
- [x] Blurring the duration field will cause a form post, that we don't want if the duration field wasn't changed. But right now, the duration value changes on change, not on blur. Do we want the same debounce logic for it perhaps?

https://community.openproject.org/projects/openproject/work_packages/41341/activity

### CODE REVIEW

File: frontend/src/app/core/days/weekday.service.ts [hunk 1]
```diff
@@ -38,13 +38,16 @@ import {
...
 import { IWeekday } from 'core-app/core/state/days/weekday.model';
-import { Observable } from 'rxjs';
+import {
+  Observable,
+  of,
+} from 'rxjs';
 
 @Injectable({ providedIn: 'root' })
 export class WeekdayService {
   @InjectField() weekdaysService:WeekdayResourceService;
```
Comment (b12f) [inline frontend/src/app/core/days/weekday.service.ts:48]: What's the difference between these two services? Can we reduce this to just one? Doesn't have to be in this PR, just wondering.

Comment (oliverguenther) [reply frontend/src/app/core/days/weekday.service.ts:48]: I think the idea was to separate akita state from business logic, but they got intermixed with loading/caching as well. the caching should happen in the akita service. I wonder what should happen with the one method depending on the weekdays to be loaded.

Do we simply mix akita state logic with business logic, or keep them separated but move the caching to the akita store?

File: frontend/src/app/shared/components/datepicker/single-date-modal/single-date.modal.ts [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+  setToday():void {
+    const today = parseDate(new Date()) as Date;
+    this.date = this.timezoneService.formattedISODate(today);
+    this.enforceManualChangesToDatepicker(today);
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  reposition(element:JQuery<HTMLElement>, target:JQuery<HTMLElement>):void {
+    element.position({
```
Comment (b12f) [inline frontend/src/app/shared/components/datepicker/single-date-modal/single-date.modal.ts:265]: Can we do this without jquery?

**Review by b12f — APPROVED** [review]
LGTM, though I'd prefer if there weren't jQuery in new code

File: frontend/src/app/shared/components/datepicker/single-date-modal/single-date.modal.ts (hunk 2 above)
Comment (oliverguenther) [reply frontend/src/app/shared/components/datepicker/single-date-modal/single-date.modal.ts:265]: Right now we're using the `jquery-ui position` helper to provide positioning with edge detection. I don't think we have an alternative library loaded right now, but in that sense its code was only moved from the previous op-datepicker-modal. Would you be okay with us adding a code maintenance ticket to look for a replacement? It's being used in these places:

https://github.com/opf/openproject/blob/dev/frontend/src/app/shared/components/op-context-menu/op-context-menu.service.ts#L116

https://github.com/opf/openproject/blob/dev/frontend/src/app/shared/components/datepicker/single-date-modal/single-date.modal.ts#L265

https://github.com/opf/openproject/blob/dev/frontend/src/app/shared/components/datepicker/multi-date-modal/multi-date.modal.ts#L367

https://github.com/opf/openproject/blob/dev/frontend/src/app/shared/components/modals/preview-modal/wp-preview-modal/wp-preview.modal.ts#L83

I'd probably have to spent some time looking for a positioning util library, or do you know any of the top of your head you can recommend?

Getting rid of jquery-ui position might allow us to remove jquery-ui altogether, although I think reporting uses it for a sortable table right now

Comment (b12f) [reply frontend/src/app/shared/components/datepicker/single-date-modal/single-date.modal.ts:265]: Works for me, I'll create a maintenance ticket. I'm not sure if we need a library for this or if one or two stateful functions that position the DOM element are enough.

**Pull request merged**